### DISCUSSION TITLE
My idiots guide for building for a Pico no longer works.

### DISCUSSION THREAD

**Discussion by beetlegigg** [micropython/micropython#13019] [RP2040 / Pico] [answered]
I had an idiots guide for building mp for a rpi pico that used to work.  By now it works no longer. First up after cloning the mp from github in a target directory my first task was to make -C mpy-cross before changing from the target directory.  This did not work but changing into the micropython/mpy-cross first and executing a 'make' seemed to work.

Im setting all this up on my new rpi5 that now runs debian bookworm (could this be why the build fails I wonder).  Anyhow this is my latest idiots guide (me being the idiot :) ) and below my guide I show the ultimate failure message.  Seems to be something to do with cmake (I told it to make not cmake??)

Idiots guide:
1) create a directory to build the repositories - e.g. pico
$ mkdir pico
$ cd pico
$ git clone https://github.com/micropython/micropython.git --branch master
2) build mpy-cross
$ cd micropython/mpy-cross
$ make 
3) go to the ports directory
$ cd /home/pi/pico/micropython/ports
4) go into the directory for the port to build - e.g. rp2
$ cd rp2
5) make the submodules for the port
$ make submodules
6) the boards directory has a list of boards that mp can be compiled for. e.g. rp2/boards
Add some links into the manifest file for links the directories containing the code to freeze into the build.  e.g. rp2/boards/RPI_PICO
7) in the build port - e.g. /home/pi/pico/micropython/ports/rp2 build the micropython.
$ make BOARD=RPI_PICO submodules
$ make BOARD=RPI_PICO clean
$ make BOARD=RPI_PICO
(not: for picoW then use RPI_PICO_W)

But the final step 7 fails and I give the output:
pi@P5-dorky:~/pico/micropython/ports/rp2 $ make BOARD=RPI_PICO submodules
make -f ../../py/mkrules.mk GIT_SUBMODULES="lib/pico-sdk" submodules
make[1]: Entering directory '/home/pi/pico/micropython/ports/rp2'
Use make V=1 or set BUILD_VERBOSE in your environment to increase build verbosity.
Updating submodules: lib/pico-sdk
Synchronizing submodule url for '../../lib/pico-sdk'
make[1]: Leaving directory '/home/pi/pico/micropython/ports/rp2'
make[1]: Entering directory '/home/pi/pico/micropython/ports/rp2'
Use make V=1 or set BUILD_VERBOSE in your environment to increase build verbosity.
Updating submodules:
make[1]: Leaving directory '/home/pi/pico/micropython/ports/rp2'

pi@P5-dorky:~/pico/micropython/ports/rp2 $ make BOARD=RPI_PICO clean
rm -f -rf build-RPI_PICO

pi@P5-dorky:~/pico/micropython/ports/rp2 $ make BOARD=RPI_PICO
[ -e build-RPI_PICO/Makefile ] || cmake -S . -B build-RPI_PICO -DPICO_BUILD_DOCS=0 -DMICROPY_BOARD=RPI_PICO -DMICROPY_BOARD_DIR=/home/pi/pico/micropython/ports/rp2/boards/RPI_PICO
/bin/sh: 1: cmake: not found
make: *** [Makefile:58: all] Error 127

As you can see some sort of error occurred with the make BOARD= RPI_PICO (I fell at the last blooming hurdle) that is probably understandable for those that are used to compiling stuff with make etc.  I hope someone can give a tip of what may be going wrong or perhaps point to some other instructions of getting a build going on Debian bookworm for a rpi pico.

**Comment by ricksorensen** [comment]
You should install `cmake` (`apt install cmake`) - which is used to build the RP2040 releases.  When you run `make` it then invokes `cmake`.  Maybe switching to the new RPi you do not have a complete install for the requirements to build RP2040 builds.   
Note: not all micropython ports need `cmake` (thank goodness in my opinion!)

**Comment by beetlegigg** [comment]
Hi @ricksorensen, thanks for the reply.  I installed cmake (sudo apt install cmake) which installed ok.  But when I now run - make BOARD=RPI_PICO I get the follow errors

pi@P5-dorky:~/pico/micropython/ports/rp2 $ make BOARD=RPI_PICO
[ -e build-RPI_PICO/Makefile ] || cmake -S . -B build-RPI_PICO -DPICO_BUILD_DOCS=0 -DMICROPY_BOARD=RPI_PICO -DMICROPY_BOARD_DIR=/home/pi/pico/micropython/ports/rp2/boards/RPI_PICO
PICO_SDK_PATH is /home/pi/pico/micropython/lib/pico-sdk
PICO platform is rp2040.
CMake Error at /home/pi/pico/micropython/lib/pico-sdk/cmake/preload/toolchains/find_compiler.cmake:28 (message):
  Compiler 'arm-none-eabi-gcc' not found, you can specify search path with
  "PICO_TOOLCHAIN_PATH".
Call Stack (most recent call first):
  /home/pi/pico/micropython/lib/pico-sdk/cmake/preload/toolchains/pico_arm_gcc.cmake:20 (pico_find_compiler)
  /usr/share/cmake-3.25/Modules/CMakeDetermineSystem.cmake:124 (include)
  CMakeLists.txt:76 (project)


CMake Error: CMake was unable to find a build program corresponding to "Unix Makefiles".  CMAKE_MAKE_PROGRAM is not set.  You probably need to select a different build tool.
CMake Error: CMAKE_C_COMPILER not set, after EnableLanguage
CMake Error: CMAKE_CXX_COMPILER not set, after EnableLanguage
-- Configuring incomplete, errors occurred!
make: *** [Makefile:58: all] Error 1

Do you, or anyone, have any thoughts on steps to take next.   I don't remember having to do anything extra on my rpi4 and Debian bullseye, but memory may be failing me. However it seems you must be correct in that the requirements for RP2040 builds are lacking somewhere.  Its a rabbit hole to be sure.

**Reply by sosi-deadeye** [reply, marked as answer]
The installation is missing the toolchain to compile code for arm.
Have you read this: https://github.com/micropython/micropython/wiki/Getting-Started#debian-ubuntu-mint-and-variants

**Comment by beetlegigg** [comment]
@sosi-deadeye,  thanks for that link I had missed that and now its working perfectly.
Apart from installing cmake (as per @ricksorensen), from the link you gave I had to install:
sudo apt-get install build-essential libreadline-dev libffi-dev git pkg-config gcc-arm-none-eabi libnewlib-arm-none-eabi
An now I'm back in business, and I've made a note of these steps in my idiots guide too.  Now I can turn in for the night and have a resplendent slumbers :-)
